python/ray/tune/tests/test_syncer.py

```diff
@@ -1006,6 +1009,147 @@ def get_remote_trial_dir(trial_id: int):
+
+    This test runs a Tune run with 3 distributed DDP workers.
+    We run 10 steps in total and checkpoint every 3 steps.
+    At the end of the test, we check the ranked index files are
+    available both locally and on the cloud.
+    We also make sure the model checkpoint files are only available
+    on the cloud.
+    """
+    exp_name = "test_dist_ckpt_to_s3"
+    local_dir = os.path.join(tmp_path, "local_dir")
+
+    def train_fn(config):
+        world_rank = session.get_world_rank()
+        for step in range(config["num_steps"]):
+            time.sleep(0.1)
+            checkpoint = None
+            if step % 3 == 0:
+                checkpoint_dir = tempfile.mkdtemp(dir=tmp_path)
+                path = os.path.join(checkpoint_dir, f"optim-{world_rank}.pt")
+                with open(path, "wb") as f:
+                    f.write(
+                        pickle.dumps(
+                            {
+                                "optimizer": "adam",
+                                "lr": 0.001,
+                                "optimizer_state": np.random.random((100, 100)),
+                            }
+                        )
+                    )
+                path = os.path.join(checkpoint_dir, f"model-{world_rank}.pt")
+                with open(path, "wb") as f:
+                    f.write(
+                        pickle.dumps(
+                            {
+                                "model": "resnet",
+                                "weights": np.random.random((100, 100)),
+                            }
+                        )
+                    )
+                checkpoint = Checkpoint.from_directory(checkpoint_dir)
+            session.report({"score": step}, checkpoint=checkpoint)
+
+    trainer = TorchTrainer(
+        train_fn,
+        train_loop_config={"num_steps": 10},
+        scaling_config=ScalingConfig(
+            num_workers=3,
+            use_gpu=False,
+        ),
+        # Note(jungong) : Trainers ignore the RunConfig specified via
+        # Tuner below. So to specify proper cloud paths and CheckpointConfig,
+        # we must pass another dummy RunConfig here.
+        # TODO(jungong) : this is extremely awkward. Refactor and clean up.
+        run_config=RunConfig(
+            storage_path=mock_s3_bucket_uri,
+            local_dir=local_dir,
+            checkpoint_config=CheckpointConfig(
+                num_to_keep=3,
+                checkpoint_frequency=3,
+                _checkpoint_keep_all_ranks=True,
+                _checkpoint_upload_from_workers=True,
+            ),
+        ),
+    )
+
+    tuner = tune.Tuner(
+        trainer,
+        run_config=RunConfig(
+            name=exp_name,
+            storage_path=mock_s3_bucket_uri,
+            local_dir=local_dir,
+        ),
+        tune_config=tune.TuneConfig(
+            # Only running 1 trial.
+            trial_dirname_creator=lambda t: "trial_0"
+        ),
+    )
+    result_grid = tuner.fit()
```

Comment:
Just to sanity check -- what is `result_grid[0].checkpoint`? I'm pretty sure it should be URI backed checkpoint pointing to the the full checkpoint in s3.

Reply:
you are right. updated the test.